ngraph/core/reference/include/ngraph/runtime/reference/proposal.hpp

```diff
@@ -0,0 +1,453 @@
+                            anchors_ptr + 2 * num_ratios * num_scales + ratio * num_scales;
+                        float* const p_anchors_hp =
+                            anchors_ptr + 3 * num_ratios * num_scales + ratio * num_scales;
+
+                        for (int scale = 0; scale < num_scales; ++scale)
+                        {
+                            // transformed width & height for given scale factors
+                            const float scale_w =
+                                0.5f * (ratio_w * scales[scale] - coordinates_offset);
+                            const float scale_h =
+                                0.5f * (ratio_h * scales[scale] - coordinates_offset);
+
+                            // (x1, y1, x2, y2) for transformed box
+                            p_anchors_wm[scale] = center - scale_w;
+                            p_anchors_hm[scale] = center - scale_h;
+                            p_anchors_wp[scale] = center + scale_w;
+                            p_anchors_hp[scale] = center + scale_h;
+
+                            if (shift_anchors)
+                            {
+                                p_anchors_wm[scale] -= half_base_size;
+                                p_anchors_hm[scale] -= half_base_size;
+                                p_anchors_wp[scale] -= half_base_size;
+                                p_anchors_hp[scale] -= half_base_size;
+                            }
+                        }
+                    }
+                    return anchors;
+                }
+
+                template <typename T>
+                static void enumerate_proposals(const T* bottom4d,
+                                                const T* d_anchor4d,
+                                                const float* anchors,
+                                                std::vector<ProposalBox<T>>& proposals,
+                                                const int num_anchors,
+                                                const int bottom_H,
+                                                const int bottom_W,
+                                                const float img_H,
+                                                const float img_W,
+                                                const float min_box_H,
+                                                const float min_box_W,
+                                                const int feat_stride,
+                                                const float box_coordinate_scale,
+                                                const float box_size_scale,
+                                                float coordinates_offset,
+                                                bool initial_clip,
+                                                bool swap_xy,
+                                                bool clip_before_nms)
+                {
+                    const int bottom_area = bottom_H * bottom_W;
```

Comment:
What if multiplication of `bottom_H` and `bottom_W` is bigger than int?

Reply:
So as we're dealing with offset here I changed the type to size_t, I think this should do OK, but if you still have doubts maybe some widening could be done.
From other point of view, we're dealing with feature map size so naturally it shouldn't hold such huge values (which doesn't mean it's not possible for such thing to happen)